contracts/PermissionLessReserve.sol

```diff
@@ -0,0 +1,596 @@
+
+    // KNC stakes
+    struct KncStakes {
+        uint128 freeKnc;    // knc that can be used to validate funds
+        uint128 kncOnStake; // per order some knc will move to be kncOnStake. part of it will be used for burning.
+    }
+    
+    //funds data
+    mapping(address => mapping(address => uint)) public makerFunds; // deposited maker funds,
+            // where order added funds are subtracted here and added to order
+    mapping(address => KncStakes) public makerKncStakes; // knc funds are required for validating deposited funds
+
+    //maker orders
+    struct FreeOrders {
+        uint32 firstOrderId;
+        uint32 numOrders; //max is 256
+        uint8 maxOrdersReached;
+        uint256 takenBitmap;
+    }
+
+    mapping(address => FreeOrders) makerOrders; //each maker will have orders that will be reused.
+
+    function PermissionLessReserve(FeeBurner burner, ERC20 knc, ERC20 token, address _admin) public {
+
+        require(knc != address(0));
+        require(token != address(0));
+        require(_admin != address(0));
+        require(burner != address(0));
+
+        feeBurnerContract = burner;
+        kncToken = knc;
+        reserveToken = token;
+        admin = _admin;
+
+        kncToken.approve(feeBurnerContract, (2**255));
+
+        //notice. if decimal API not supported this should revert
+        setDecimals(reserveToken);
+        require(getDecimals(reserveToken) <= MAX_DECIMALS);
+    }
+
+    function getConversionRate(ERC20 src, ERC20 dest, uint totalSrcAmount, uint blockNumber) public view returns(uint) {
+
+        require((src == ETH_TOKEN_ADDRESS) || (dest == ETH_TOKEN_ADDRESS));
+        require((src == reserveToken) || (dest == reserveToken));
+        blockNumber; // in this reserve no order expiry == no use for blockNumber. here to avoid compiler warning.
+
+        uint32 orderId;
+
+        if (src == ETH_TOKEN_ADDRESS) {
+            if (isNextOrderTail(BUY_HEAD_ID)) return 0;
+            orderId = BUY_HEAD_ID;
+        } else {
+            if (isNextOrderTail(SELL_HEAD_ID)) return 0;
+            orderId = SELL_HEAD_ID;
+        }
+
+        uint128 remainingSrcAmount = uint128(totalSrcAmount);
+        uint128 totalDstAmount = 0;
+
+        while (!isNextOrderTail(orderId)) {
+
+            orderId = getNextOrderId(orderId);
+
+            Order memory order = orders[orderId];
+
+            if (order.srcAmount < remainingSrcAmount) {
+                totalDstAmount += order.dstAmount;
+                remainingSrcAmount -= order.srcAmount;
+            } else {
+                totalDstAmount += order.dstAmount * remainingSrcAmount / order.srcAmount;
+                remainingSrcAmount = 0;
+                break;
+            }
+        }
+
+        if ((remainingSrcAmount != 0) || (totalDstAmount == 0)) return 0; //not enough tokens to exchange.
+
+        //check overflow
+        if (uint(totalDstAmount) * PRECISION < uint(totalDstAmount)) return 0;
```

Comment:
you probably want to check that total dest amount < max qty.

Reply:
beside, something else is wrong here. is totaldestamount is 128 bits, you will never have overflow.